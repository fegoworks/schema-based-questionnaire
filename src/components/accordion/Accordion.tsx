import "./accordion.scss";
import { FunctionComponent, ReactElement, ReactNode, useState } from "react";

interface AccordionProps {
  children?: ReactElement;
  className?: string;
  title?: string | ReactNode;
  answer?: string | string[] | number;
  isExpanded: boolean;
  toggleExpanded: () => void;
  isStatusGreen: boolean;
}
const Accordion: FunctionComponent<AccordionProps> = ({
  children,
  title,
  answer,
  isExpanded,
  toggleExpanded,
  isStatusGreen,
}) => {
  return (
    <div className="Accordion">
      <div className={"accordion-item"} onClick={toggleExpanded}>
        <div
          className={`accordion-item-header ${
            isExpanded ? "accordion-expanded" : ""
          }`}>
          <span>{title}</span>
          {!isExpanded && (
            <>
              <span>{answer}</span>
              <span
                className={`status ${
                  isStatusGreen ? "status-green" : ""
                }`}></span>
            </>
          )}
        </div>
      </div>
      <div
        className={`accordion-body ${isExpanded ? "accordion-expanded" : ""}`}>
        {children}
      </div>
    </div>
  );
};

export default Accordion;
