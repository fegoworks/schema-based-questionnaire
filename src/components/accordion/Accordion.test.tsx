import { fireEvent, render, screen } from "@testing-library/react";
import Accordion from "./Accordion";

const setup = () => {
  const defaultProps = {
    title: "Accordion Title",
    answer: "Accordion Answer",
    isExpanded: false,
    toggleExpanded: () => {},
    isStatusGreen: false,
  };

  const openedStateProps = { ...defaultProps, isExpanded: true };

  return {
    defaultProps,
    openedStateProps,
  };
};

describe("Component - Accordion", () => {
  it("should render accordion component", () => {
    const { defaultProps } = setup();
    const { container } = render(
      <Accordion {...defaultProps}>
        <div>Accordion content</div>
      </Accordion>
    );
    expect(container).toMatchSnapshot();
  });

  it("should show title and answer in closed state", async () => {
    const { defaultProps } = setup();
    const { getByText } = render(
      <Accordion {...defaultProps}>
        <div>Accordion content</div>
      </Accordion>
    );
    expect(getByText("Accordion Title")).toBeTruthy();
    expect(getByText("Accordion Answer")).toBeTruthy();
  });

  it("should open accordion on click", async () => {
    const { defaultProps } = setup();
    const { container } = render(
      <Accordion {...defaultProps}>
        <div>Accordion content</div>
      </Accordion>
    );
    fireEvent.click(container);
    expect(await screen.findByText(/Accordion content/i)).toBeInTheDocument();
  });

  it("should render only accordion answer in header when in open state", async () => {
    const { openedStateProps } = setup();
    const { getByText } = render(
      <Accordion {...openedStateProps}>
        <>Accordion content</>
      </Accordion>
    );
    const answerComponent = screen.queryByText("Accordion Answer");
    expect(getByText("Accordion content")).toHaveClass("accordion-expanded");
    expect(answerComponent).not.toBeInTheDocument();
  });
});
