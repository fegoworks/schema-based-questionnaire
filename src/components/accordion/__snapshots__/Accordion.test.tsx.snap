// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Component - Accordion > should render accordion component 1`] = `
<div>
  <div
    class="Accordion"
  >
    <div
      class="accordion-item"
    >
      <div
        class="accordion-item-header "
      >
        <span>
          Accordion Title
        </span>
        <span>
          Accordion Answer
        </span>
        <span
          class="status "
        />
      </div>
    </div>
    <div
      class="accordion-body "
    >
      <div>
        Accordion content
      </div>
    </div>
  </div>
</div>
`;
